susemanager-utils/susemanager-sls/src/modules/mgr_caasp_manager.py

```diff
@@ -175,19 +229,114 @@ def add_node(skuba_cluster_path,
     return ret
 
 
+def _upgrade_cluster_plan(skuba_cluster_path,
+                          verbosity=None,
+                          timeout=DEFAULT_TIMEOUT,
+                          **kwargs):
+
+    cmd_args = "cluster upgrade plan"
+
+    if verbosity:
+        cmd_args += " --verbosity {}".format(verbosity)
+
+    skuba_proc = _call_skuba(skuba_cluster_path, cmd_args, timeout=timeout)
+    if skuba_proc.process.returncode != 0:
+        error_msg = "Unexpected error {} at skuba when upgrading the cluster: {}".format(
+                skuba_proc.process.returncode,
+                salt.utils.stringutils.to_str(skuba_proc.stderr))
+        log.error(error_msg)
+
+    ret = {
+        'stdout': salt.utils.stringutils.to_str(skuba_proc.stdout),
+        'stderr': salt.utils.stringutils.to_str(skuba_proc.stderr),
+        'success': not skuba_proc.process.returncode,
+        'retcode': skuba_proc.process.returncode,
+    }
+    return ret
+
+
 def upgrade_cluster(skuba_cluster_path,
                     verbosity=None,
                     timeout=DEFAULT_TIMEOUT,
+                    plan=False,
                     **kwargs):
 
-    cmd_args = "cluster upgrade plan"
+    if plan:
+        return _upgrade_cluster_plan(skuba_cluster_path=skuba_cluster_path,
+                                     verbosity=verbosity,
+                                     timeout=timeout,
+                                     **kwargs)
+
+    # Perform the cluster upgrade procedure.
+    # 1. Upgrade addons
+    # 2. Upgrade all nodes
+    # 3. Upgrade addons
+    ret = {
+        'success' : True,
+        'retcode' : 0,
+        'stage0_upgrade_addons': {},
+        'stage1_upgrade_nodes': {},
+        'stage2_upgrade_addons': {},
+    }
+
+    ret['stage0_upgrade_addons'] = upgrade_addons(skuba_cluster_path=skuba_cluster_path,
+                                                  verbosity=verbosity,
+                                                  timeout=timeout,
+                                                  plan=plan,
+                                                  **kwargs)
+
+    if not ret['stage0_upgrade_addons']['success']:
+        ret['success'] = False
+        return ret
+
+    nodes = list_nodes(skuba_cluster_path=skuba_cluster_path,
```

Comment:
From what I see there's no safety in that `list_nodes` will list control plane nodes first, right? I think I have always seen them listed first, but it would be good if we ensure that this is the case, so when we upgrade nodes we enforce the upgrade of the control plane nodes first, then the workers.

Reply:
Thanks for the feedback @ereslibre.

I've push a minor change to ensure master nodes are always upgraded first :+1: